README.md
# <img src="assets/favicon.png" alt="icon" width="4%"/> tor-brute - Forms brute-forcer via Tor

[![GitHub license](https://img.shields.io/github/license/ThomasByr/tor-brute)](https://github.com/ThomasByr/tor-brute/blob/master/LICENSE)
[![GitHub commits](https://badgen.net/github/commits/ThomasByr/tor-brute)](https://GitHub.com/ThomasByr/tor-brute/commit/)
[![GitHub latest commit](https://badgen.net/github/last-commit/ThomasByr/tor-brute)](https://gitHub.com/ThomasByr/tor-brute/commit/)
[![Maintenance](https://img.shields.io/badge/maintained%3F-yes-green.svg)](https://GitHub.com/ThomasByr/tor-brute/graphs/commit-activity)

[![Python Package](https://github.com/ThomasByr/tor-brute/actions/workflows/python-package.yml/badge.svg)](https://github.com/ThomasByr/tor-brute/actions/workflows/python-package.yml)
[![GitHub release](https://img.shields.io/github/release/ThomasByr/tor-brute)](https://github.com/ThomasByr/tor-brute/releases/)
[![Author](https://img.shields.io/badge/author-@ThomasByr-blue)](https://github.com/ThomasByr)
[![Author](https://img.shields.io/badge/author-@ThomasD-blue)](https://github.com/LosKeeper)

1. [✏️ Setup](#️-setup)
2. [🔧 Usage](#-usage)
3. [🧑‍🏫 Contributing](#-contributing)
4. [⚖️ License](#️-license)
5. [🔄 Changelog](#-changelog)
6. [🐛 Bugs and TODO](#-bugs-and-todo)
7. [🎨 Logo and Icons](#-logo-and-icons)

> <picture>
>   <source media="(prefers-color-scheme: light)" srcset="https://raw.githubusercontent.com/Mqxx/GitHub-Markdown/main/blockquotes/badge/light-theme/warning.svg">
>   <img alt="Warning" src="https://raw.githubusercontent.com/Mqxx/GitHub-Markdown/main/blockquotes/badge/dark-theme/warning.svg">
> </picture><br>
>
> Disclaimer : This is an educational project, please do not use it for illegal purposes. We are not responsible for any damage caused by the use of this project.

## ✏️ Setup

> <picture>
>   <source media="(prefers-color-scheme: light)" srcset="https://raw.githubusercontent.com/Mqxx/GitHub-Markdown/main/blockquotes/badge/light-theme/info.svg">
>   <img alt="Info" src="https://raw.githubusercontent.com/Mqxx/GitHub-Markdown/main/blockquotes/badge/dark-theme/info.svg">
> </picture><br>
>
> Please note we do not officially support Windows or MacOS, but we do provide some instructions for those who want to use it on these platforms.

You do not explicitly need a conda environment for the bot to run. But it is always recommended nontheless, especially because the next LTS of Ubuntu won't let users pip-install anything without a virtual environment. At the time of writing, this app requires `python >= 3.8` to run.

First, install the dependencies :

```bash
sudo apt-get update
sudo apt-get install tor
sudo service tor stop  # the app will start its own tor instance
# or
# sudo systemctl disable tor
```

Then clone the repository and cd into it :

```bash
# Clones the repository
git clone git@github.com:ThomasByr/tor-brute.git
cd tor-brute
```

You can create and activate a conda environment with the following commands :

```bash
# Creates environment and install dependencies
conda env create -f environment.yml -y
conda activate tor
```

Finally, run the app in the background with `nohup` :

```bash
# Runs the app in the background
nohup python tor_brute.py &
```

or in the foreground :

```bash
# Runs the app (lets you Ctrl+C to stop it)
python tor_brute.py
```

## 🔧 Usage

Simply create a `.cfg` file from [.cfg.example](.cfg.example) and fill it, then provide text files for both usernames and passwords. The app will try every combination of usernames and passwords, and will issue a log record for each successful login. If you have a user file that looks like this :

```txt
foo
bar
baz
```

The app will try these usernames (for combinations of size 2) (note that if you use `--all`, the app will create permutations instead of combinations) :

```txt
foo
bar
baz
foobar
foobaz
barbaz
```

and for each one, try every password combination in the password file following the same logic.

Once you are all set, you can run the app with :

<!-- markdownlint-disable MD051 -->

| argument                  | hint                                       | default             |
| ------------------------- | ------------------------------------------ | ------------------- |
| `-h`, `--help`            | show help message and exit                 |                     |
| `-v`, `--version`         | show program's version                     |                     |
| `-d`, `--debug`           | debug mode                                 | `False`             |
| `-c`, `--config`          | path to the config file                    | `.cfg`              |
| `-u`, `--user`            | path to the usernames file                 | `assets/user.txt`   |
| `-p`, `--passwd`          | path to the passwords file                 | `assets/passwd.txt` |
| `-i`, `--iter` [\*][1]    | number of combination for user/passwd      | `3, 2`              |
| `-e`, `--each`            | change Tor ID each X requests (0 or >=100) | `1000`              |
| `-t`, `--timeout`         | timeout for http requests                  | `10`                |
| `-m`, `--max-tries`       | how much to _retry_ stuff                  | `3`                 |
| `-w`, `--threads` [\*][2] | number of threads                          | `50`                |
| `-s`, `--sleep`           | additional sleep time between each ID swap | `0`                 |
| `-a`, `--all`             | use permutations (unordered combination)   | `False`             |

[1]: ## "a file with a, b, c with iter=2 would produce a, b, c, ab, ac, ba, bc, ca, cb"
[2]: ## "too many threads might result in a server timeout"

<!-- markdownlint-enable MD051 -->

## 🧑‍🏫 Contributing

If you ever want to contribute, either request the contributor status, or, more manually, fork the repo and make a pull request !

We are using [black](https://github.com/psf/black) to format the code, so make sure you have it installed and run :

```ps1
black src
```

> The standard procedure is :
>
> ```txt
> fork -> git branch -> push -> pull request
> ```
>
> Note that we won't accept any PR :
>
> - that does not follow our Contributing Guidelines
> - that is not sufficiently commented or isn't well formated
> - without any proper test suite
> - with a failing or incomplete test suite

Happy coding ! 🙂

## ⚖️ License

> <picture>
>   <source media="(prefers-color-scheme: light)" srcset="https://raw.githubusercontent.com/Mqxx/GitHub-Markdown/main/blockquotes/badge/light-theme/warning.svg">
>   <img alt="Warning" src="https://raw.githubusercontent.com/Mqxx/GitHub-Markdown/main/blockquotes/badge/dark-theme/warning.svg">
> </picture><br>
>
> This repository is an app and might be exported as a standalone binary. Working source code is licensed under AGPL, the text assets are unlicensed, images and logos have their own section. The following license only applies to the template itself and is not legal advice. <FONT COLOR="#ff0000"><u>The license of this repo does not apply to the resources used in it.</u></FONT> Please check the license of each resource before using them.

This project is licensed under the AGPL-3.0 new or revised license. Please read the [LICENSE](LICENSE.md) file. Additionally :

- Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

- Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

- Neither the name of the tor-brute authors nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

```LICENSE
tor-brute - Forms brute-forcer via Tor
Copyright (C) 2023 ThomasByr AND LosKeeper

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.
```

## 🔄 Changelog

Please read the [changelog](changelog.md) file for the full history !

<details>
  <summary>  first release (click here to expand) </summary>

**v1.0** beta candidate (1.0.0-dev)

- add option to change Tor ID each X requests
- new `TupleGenerator` that yields products of combinations
- renew http session each Tor ID swap
- `ThreadPool` is not closed/joined/terminated/deleted and then recreated anymore ! we use POSIX condition variables !
- somehow improved performance by 6.9% (not sure how)
- RAM usage does not seem to increase anymore (to be confirmed)
- consistent naming for variables and files
- next up: beta, release candidate, and release (drastic changes should only happen between beta and release candidate)

**v1.0** beta (1.0.0-beta.1 and 1.0.1-beta.1)

- `-t` for timeout, the maximum number of seconds to wait for one request
- `-m` for max retries, the maximum number of retries for one request, as well as the maximum number of consecutive failures before shutting down
- `-w` for workers of threads, pretty self-explanatory
- `-s` for sleep, the amount of seconds to wait between each Tor ID swap
- `-a` for use_all or all, to use permutations instead of combinations in generators
- each worker has its own session (no more shared session) and is renewed each Tor ID swap

**v1.0** candidate (1.0.1-rc1)

- no breaking changes here
- few bug fixes and minor refactors
- opened Tor proxy to http and https (this should not slow down the process)
- faster thread identification (no more `threading.current_thread().name.split('-')[1].split(' ')[0]`)

**v1.0** release (1.0.1)

- config file checkers
- tests for generators
- generators now skip empty or blank lines in .txt files and do no longer lowercase everything
- fixed consecutive failure logic and logging
- changed emotes ... because

**v1.1** (1.1.0)

- removed `no-tor` suggestion

</details>

## 🐛 Bugs and TODO

**TODO** (first implementation version)

- [x] add a simple cli (0.1.1)
- [x] option to change Tor ID each X requests (would need to implement a catch-up mechanism because thread jobs are unordered) (v1.0.0-dev)
- [ ] option to use a running tor instance/service
- [ ] choose protocol (http, https, ssh, etc.) (opened http and https in 1.0.1-rc1)
- [x] dynamic change between `combinations` and `permutations` (v1.0.0-beta.1)

**Known Bugs** (latest fix)

- [ ] lagging threads are not catching up, especially when `ReadTimeout` is reached (interferes with Tor ID swap)
- [x] sometimes, successfull logins are not reported, or are reported twice (v1.0.1)

## 🎨 Logo and Icons

Icon by Dewi Sari from [flaticon.com](https://www.flaticon.com/fr/auteurs/dewi-sari)

Unless otherwise stated, all icons and logos are made by the author.
Copyright (C) 2023 Thomas BOUYER, all rights reserved.

Tools used :

- [Microsoft Designer](https://designer.microsoft.com/)
- [Clip Studio Paint](https://www.clipstudio.net/en)
- [Canva](https://www.canva.com/)


CHANGELOG.md
# Changelog

<summary>The full history, or so was I told...</summary>
versions next to titles correspond to the first version that introduced either the release or the minor

## alpha

**v0.1** first public release

- use threads
- create a cli
- unique progression bar for all users (so this is not spammy)
- removed color and emotes from streams that don't support it

**v0.2** misc refactors

- use `argparse` instead of `sys.argv`
- better formula to count combinations
- beautify code and error messages/handling
- account for `ReadTimeout` in `requests` (and other errors)
- account for ip check failure, successive failures, and exit node unchanged
- `logging.critical` sends SIGTERM to all threads and shows cursor back

## first release

**v1.0** beta candidate (1.0.0-dev)

- add option to change Tor ID each X requests
- new `TupleGenerator` that yields products of combinations
- renew http session each Tor ID swap
- `ThreadPool` is not closed/joined/terminated/deleted and then recreated anymore ! we use POSIX condition variables !
- somehow improved performance by 6.9% (not sure how)
- RAM usage does not seem to increase anymore (to be confirmed)
- consistent naming for variables and files
- next up: beta, release candidate, and release (drastic changes should only happen between beta and release candidate)

**v1.0** beta (1.0.0-beta.1 and 1.0.1-beta.1)

- `-t` for timeout, the maximum number of seconds to wait for one request
- `-m` for max retries, the maximum number of retries for one request, as well as the maximum number of consecutive failures before shutting down
- `-w` for workers of threads, pretty self-explanatory
- `-s` for sleep, the amount of seconds to wait between each Tor ID swap
- `-a` for use_all or all, to use permutations instead of combinations in generators
- each worker has its own session (no more shared session) and is renewed each Tor ID swap

**v1.0** candidate (1.0.1-rc1)

- no breaking changes here
- few bug fixes and minor refactors
- opened Tor proxy to http and https (this should not slow down the process)
- faster thread identification (no more `threading.current_thread().name.split('-')[1].split(' ')[0]`)

**v1.0** release (1.0.1)

- config file checkers
- tests for generators
- generators now skip empty or blank lines in .txt files and do no longer lowercase everything
- fixed consecutive failure logic and logging
- changed emotes ... because

**v1.1** (1.1.0)

- removed `no-tor` suggestion


src/core/cli_parser.py
from argparse import ArgumentParser, Namespace
from dataclasses import dataclass
from typing import Any

from typing_extensions import override

from ..version import __version__

__all__ = ["Args", "parser", "check_args"]


@dataclass
class Args:
    cfg_path: str = ".cfg"
    debug: bool = False
    usernames: str = "assets/user.txt"
    passwords: str = "assets/passwd.txt"
    it_comb: tuple[int, int] = (3, 2)
    each: int = 1000
    timeout: int = 10
    max_tries: int = 3
    threads: int = 50
    sleep: int = 0
    use_all: bool = False


class WeakParser(ArgumentParser):
    @override
    def add_argument(self, *args: Any, **kwargs: Any) -> "WeakParser":
        kwargs["default"] = None
        super().add_argument(*args, **kwargs)
        return self

    def add_path_argument(self, *args: Any, **kwargs: Any) -> "WeakParser":
        kwargs["metavar"] = "P"
        kwargs["type"] = str
        super().add_argument(*args, **kwargs)
        return self

    def add_int_argument(self, *args: Any, **kwargs: Any) -> "WeakParser":
        kwargs["metavar"] = "N"
        kwargs["type"] = int
        super().add_argument(*args, **kwargs)
        return self


def parse_it_comb(it_comb: str) -> tuple[int, int]:
    # parse from "1, 2" to (1, 2)
    # parse from "1" to (1, 1)
    parts = list(map(lambda s: s.strip(), it_comb.split(",")))
    if len(parts) == 1:
        return (int(parts[0]), int(parts[0]))
    elif len(parts) == 2:
        return (int(parts[0]), int(parts[1]))
    return  # leave blank since argparse does not reraise


def parser() -> WeakParser:
    parser = WeakParser(
        description="A simple CLI to bruteforce login forms",
        epilog="visit us on GitHub : https://github.com/ThomasByr/tor-brute",
    )
    return (
        parser.add_argument("-v", "--version", action="version", version=f"%(prog)s {__version__}")
        .add_path_argument("-c", "--config", dest="cfg_path", help="Path to the config file")
        .add_path_argument(
            "-u",
            "--user",
            dest="usernames",
            help="Path to the usernames text file (one username-part per line))",
        )
        .add_path_argument(
            "-p",
            "--passwd",
            dest="passwords",
            help="Path to the passwords text file (one password-part per line)",
        )
        .add_argument(
            "-i",
            "--iter",
            dest="it_comb",
            metavar="N[, M]",
            type=parse_it_comb,
            help="how much to iterate over the combinations (default: 3, 2 for user, passwd)",
        )
        .add_int_argument(
            "-e",
            "--each",
            dest="each",
            help="each how many requests to change the Tor identity (default: 1000)",
        )
        .add_int_argument(
            "-t",
            "--timeout",
            dest="timeout",
            help="how much to wait for a response in seconds (default: 10)",
        )
        .add_int_argument(
            "-m",
            "--max-tries",
            dest="max_tries",
            help="how much to try to reconnect to the target and how many consecutive fail to consider before killing the app (default: 3)",
        )
        .add_int_argument("-w", "--threads", dest="threads", help="how much threads to use (default: 50)")
        .add_int_argument(
            "-s",
            "--sleep",
            dest="sleep",
            help="how much to sleep between each Tor identity swap (default: 0)",
        )
        .add_argument(
            "-a",
            "--all",
            dest="use_all",
            action="store_true",
            help="use all the possible permutations instead of the faster unordered classic (default: False)",
        )
        .add_argument("-d", "--debug", dest="debug", action="store_true", help="Enable debug mode")
    )


def check_path(path: str, mode: str = "r") -> str:
    """Check if the path exists"""
    try:
        with open(path, mode, encoding="utf-8"):
            return path
    except FileNotFoundError as e:
        raise ValueError(f"file {path} does not exist") from e
    except PermissionError as e:
        raise ValueError(f"file {path} is not can not be opened in [{mode}] mode") from e
    except IsADirectoryError as e:
        raise ValueError(f"file {path} is a directory") from e
    except OSError as e:
        raise ValueError(f"file {path} is not a file") from e
    except Exception as e:
        raise ValueError(f"unknown error with file {path}") from e


def check_args(args: Namespace) -> Args:
    MINIMUM_EACH_VALUE = 100
    a = Args()

    # paths (and read property)
    if args.cfg_path is not None:
        a.cfg_path = check_path(args.cfg_path)
    if args.usernames is not None:
        a.usernames = check_path(args.usernames)
    if args.passwords is not None:
        a.passwords = check_path(args.passwords)

    if args.it_comb is not None:
        if any(i < 1 for i in args.it_comb):
            raise ValueError("[--iter] values must be >= 1")
        a.it_comb = args.it_comb
    if args.each is not None:
        if 0 < args.each < MINIMUM_EACH_VALUE:
            raise ValueError("[--each] value must be at least >= %d", MINIMUM_EACH_VALUE)
        a.each = args.each
    if args.timeout is not None:
        if args.timeout < 1:
            raise ValueError("[--timeout] value must be >= 1")
        a.timeout = args.timeout
    if args.max_tries is not None:
        if args.max_tries < 1:
            raise ValueError("[--max-tries] value must be >= 1")
        a.max_tries = args.max_tries
    if args.threads is not None:
        if args.threads < 1:
            raise ValueError("[--threads] value must be >= 1")
        a.threads = args.threads
    if args.sleep is not None:
        if args.sleep < 0:
            raise ValueError("[--sleep] value must be >= 0")
        a.sleep = args.sleep
    if args.use_all is not None:
        a.use_all = args.use_all
    if args.debug:
        a.debug = args.debug
    return a
